File: shared/engine/table.c
/**
 * @file table.c
 * @brief Logic Engine - Table Lookup Implementation
 */

#include "table.h"

/*============================================================================
 * Helper Functions
 *============================================================================*/

uint8_t Table_FindIndex(const int32_t* arr, uint8_t count, int32_t value)
{
    if (count == 0) return 0;
    if (count == 1) return 0;

    /* Value below first point */
    if (value <= arr[0]) return 0;

    /* Value above last point */
    if (value >= arr[count - 1]) return count - 2;

    /* Binary search for interval */
    uint8_t low = 0;
    uint8_t high = count - 1;

    while (low < high - 1) {
        uint8_t mid = (low + high) / 2;
        if (value < arr[mid]) {
            high = mid;
        } else {
            low = mid;
        }
    }

    return low;
}

int32_t Table_Lerp(int32_t x, int32_t x0, int32_t x1, int32_t y0, int32_t y1)
{
    /* Avoid division by zero */
    int32_t dx = x1 - x0;
    if (dx == 0) return y0;

    /* Calculate interpolation using 64-bit to avoid overflow */
    int64_t dy = (int64_t)(y1 - y0);
    int64_t offset = (int64_t)(x - x0);

    return y0 + (int32_t)((dy * offset) / dx);
}

/*============================================================================
 * 2D Table Functions
 *============================================================================*/

int32_t Table2D_Lookup(const Table2D_t* table, int32_t x)
{
    if (!table || table->count == 0) return 0;

    /* Single point table */
    if (table->count == 1) return table->y[0];

    /* Find interval */
    uint8_t i = Table_FindIndex(table->x, table->count, x);

    /* Clamp to valid range for interpolation */
    if (i >= table->count - 1) i = table->count - 2;

    /* Linear interpolation */
    return Table_Lerp(x,
                      table->x[i], table->x[i + 1],
                      table->y[i], table->y[i + 1]);
}

int32_t Table2D_LookupRaw(const Table2D_t* table, int32_t x)
{
    if (!table || table->count == 0) return 0;

    uint8_t i = Table_FindIndex(table->x, table->count, x);
    return table->y[i];
}

int32_t Table2D_Validate(const Table2D_t* table)
{
    if (!table) return 0;
    if (table->count == 0 || table->count > TABLE_2D_MAX_POINTS) return 0;

    /* Check X values are sorted */
    for (uint8_t i = 1; i < table->count; i++) {
        if (table->x[i] <= table->x[i - 1]) return 0;
    }

    return 1;
}

/*============================================================================
 * 3D Table Functions
 *============================================================================*/

int32_t Table3D_Lookup(const Table3D_t* table, int32_t x, int32_t y)
{
    if (!table || table->x_count == 0 || table->y_count == 0) return 0;

    /* Single cell table */
    if (table->x_count == 1 && table->y_count == 1) {
        return table->z[0][0];
    }

    /* Find X interval */
    uint8_t xi = 0;
    if (table->x_count > 1) {
        xi = Table_FindIndex(table->x, table->x_count, x);
        if (xi >= table->x_count - 1) xi = table->x_count - 2;
    }

    /* Find Y interval */
    uint8_t yi = 0;
    if (table->y_count > 1) {
        yi = Table_FindIndex(table->y, table->y_count, y);
        if (yi >= table->y_count - 1) yi = table->y_count - 2;
    }

    /* Handle edge cases */
    if (table->x_count == 1) {
        /* 1D interpolation along Y only */
        return Table_Lerp(y,
                          table->y[yi], table->y[yi + 1],
                          table->z[yi][0], table->z[yi + 1][0]);
    }

    if (table->y_count == 1) {
        /* 1D interpolation along X only */
        return Table_Lerp(x,
                          table->x[xi], table->x[xi + 1],
                          table->z[0][xi], table->z[0][xi + 1]);
    }

    /* Bilinear interpolation */
    /* Get corner values */
    int32_t z00 = table->z[yi][xi];         /* Bottom-left */
    int32_t z10 = table->z[yi][xi + 1];     /* Bottom-right */
    int32_t z01 = table->z[yi + 1][xi];     /* Top-left */
    int32_t z11 = table->z[yi + 1][xi + 1]; /* Top-right */

    /* Get axis values */
    int32_t x0 = table->x[xi];
    int32_t x1 = table->x[xi + 1];
    int32_t y0 = table->y[yi];
    int32_t y1 = table->y[yi + 1];

    /* Interpolate along X for both Y rows */
    int32_t z_y0 = Table_Lerp(x, x0, x1, z00, z10);
    int32_t z_y1 = Table_Lerp(x, x0, x1, z01, z11);

    /* Interpolate along Y */
    return Table_Lerp(y, y0, y1, z_y0, z_y1);
}

int32_t Table3D_LookupRaw(const Table3D_t* table, int32_t x, int32_t y)
{
    if (!table || table->x_count == 0 || table->y_count == 0) return 0;

    uint8_t xi = Table_FindIndex(table->x, table->x_count, x);
    uint8_t yi = Table_FindIndex(table->y, table->y_count, y);

    return table->z[yi][xi];
}

int32_t Table3D_Validate(const Table3D_t* table)
{
    if (!table) return 0;
    if (table->x_count == 0 || table->x_count > TABLE_3D_MAX_X) return 0;
    if (table->y_count == 0 || table->y_count > TABLE_3D_MAX_Y) return 0;

    /* Check X values are sorted */
    for (uint8_t i = 1; i < table->x_count; i++) {
        if (table->x[i] <= table->x[i - 1]) return 0;
    }

    /* Check Y values are sorted */
    for (uint8_t i = 1; i < table->y_count; i++) {
        if (table->y[i] <= table->y[i - 1]) return 0;
    }

    return 1;
}

File: shared/python/engine/table.py
"""
Logic Engine - Lookup Tables

2D and 3D lookup tables with linear/bilinear interpolation.
"""

from dataclasses import dataclass, field
from typing import List, Optional


TABLE_2D_MAX_POINTS = 16
TABLE_3D_MAX_X = 8
TABLE_3D_MAX_Y = 8


@dataclass
class Table2D:
    """2D lookup table with interpolation"""
    x_values: List[int] = field(default_factory=list)
    y_values: List[int] = field(default_factory=list)

    @property
    def size(self) -> int:
        return min(len(self.x_values), len(self.y_values))


@dataclass
class Table3D:
    """3D lookup table with bilinear interpolation"""
    x_values: List[int] = field(default_factory=list)
    y_values: List[int] = field(default_factory=list)
    z_values: List[List[int]] = field(default_factory=list)  # z[y][x]

    @property
    def x_size(self) -> int:
        return len(self.x_values)

    @property
    def y_size(self) -> int:
        return len(self.y_values)


def _find_index(values: List[int], x: int) -> tuple[int, int]:
    """
    Binary search to find indices for interpolation.
    Returns (lower_index, upper_index).
    """
    n = len(values)
    if n == 0:
        return 0, 0

    if x <= values[0]:
        return 0, 0
    if x >= values[n - 1]:
        return n - 1, n - 1

    # Binary search
    low = 0
    high = n - 1

    while high - low > 1:
        mid = (low + high) // 2
        if values[mid] <= x:
            low = mid
        else:
            high = mid

    return low, high


def _interpolate(x: int, x0: int, x1: int, y0: int, y1: int) -> int:
    """Linear interpolation with integer math"""
    if x1 == x0:
        return y0

    # Use 64-bit intermediate to prevent overflow
    dx = x1 - x0
    dy = y1 - y0
    result = y0 + (x - x0) * dy // dx

    return result


def table2d_lookup(table: Table2D, x: int) -> int:
    """
    Lookup value in 2D table with linear interpolation.

    Args:
        table: 2D lookup table
        x: Input value

    Returns:
        Interpolated output value
    """
    n = table.size
    if n == 0:
        return 0
    if n == 1:
        return table.y_values[0]

    i0, i1 = _find_index(table.x_values, x)

    if i0 == i1:
        return table.y_values[i0]

    return _interpolate(
        x,
        table.x_values[i0], table.x_values[i1],
        table.y_values[i0], table.y_values[i1]
    )


def table3d_lookup(table: Table3D, x: int, y: int) -> int:
    """
    Lookup value in 3D table with bilinear interpolation.

    Args:
        table: 3D lookup table
        x: X input value
        y: Y input value

    Returns:
        Interpolated output value
    """
    nx = table.x_size
    ny = table.y_size

    if nx == 0 or ny == 0 or not table.z_values:
        return 0

    # Find x indices
    xi0, xi1 = _find_index(table.x_values, x)

    # Find y indices
    yi0, yi1 = _find_index(table.y_values, y)

    # Get corner values
    try:
        z00 = table.z_values[yi0][xi0]
        z01 = table.z_values[yi0][xi1] if xi1 < nx else z00
        z10 = table.z_values[yi1][xi0] if yi1 < ny else z00
        z11 = table.z_values[yi1][xi1] if (xi1 < nx and yi1 < ny) else z00
    except IndexError:
        return 0

    # Bilinear interpolation
    if xi0 == xi1:
        # Only Y interpolation needed
        if yi0 == yi1:
            return z00
        return _interpolate(y, table.y_values[yi0], table.y_values[yi1], z00, z10)

    if yi0 == yi1:
        # Only X interpolation needed
        return _interpolate(x, table.x_values[xi0], table.x_values[xi1], z00, z01)

    # Full bilinear interpolation
    # Interpolate along X for both Y rows
    zx0 = _interpolate(x, table.x_values[xi0], table.x_values[xi1], z00, z01)
    zx1 = _interpolate(x, table.x_values[xi0], table.x_values[xi1], z10, z11)

    # Interpolate along Y
    return _interpolate(y, table.y_values[yi0], table.y_values[yi1], zx0, zx1)


def table2d_from_points(points: List[tuple[int, int]]) -> Table2D:
    """Create 2D table from list of (x, y) points"""
    points = sorted(points, key=lambda p: p[0])
    return Table2D(
        x_values=[p[0] for p in points],
        y_values=[p[1] for p in points]
    )
